scipy/interpolate/interpolate.py

```diff
@@ -75,19 +84,26 @@ def lagrange(x, w):
     ...          label=r"$3 x^2 - 2 x$", linestyle='-.')
     >>> plt.legend()
     >>> plt.show()
-
     """
 
     M = len(x)
     p = poly1d(0.0)
     for j in range(M):
-        pt = poly1d(w[j])
+        pt = poly1d(y[j])
         for k in range(M):
             if k == j:
                 continue
             fac = x[j]-x[k]
             pt *= poly1d([1.0, -x[k]])/fac
         p += pt
+    diff = M - len(p.coeffs)
+    if reduce is False:
+        if diff > 0:
+            tmp = np.zeros(diff)
+            coef = np.append(tmp, p.coeffs)
+            return p, coef
```

Comment:
This all needs a comment. The logic is not hard to parse when you look at it, but the first glance is puzzling, so it needs a short comment.